core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java

```diff
@@ -385,29 +405,427 @@ public synchronized void close()
+            writeBuffer = null;
+            try (Closer closer = Closer.create()) {
+                closer.register(exchange);
+                if (exchangeSink != null) {
+                    closer.register(exchangeSink::abort);
+                }
+            }
+            catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+    }
+
+    @NotThreadSafe
+    private interface OutputSource
+            extends Closeable
+    {
+        Slice getNext();
+
+        boolean isFinished();
+
+        ListenableFuture<Void> isBlocked();
+
+        long getRetainedSizeInBytes();
+    }
+
+    @NotThreadSafe
+    private static class InMemoryBufferOutputSource
+            implements OutputSource
+    {
+        private final Iterator<Slice> iterator;
+
+        private InMemoryBufferOutputSource(Iterator<Slice> iterator)
+        {
+            this.iterator = requireNonNull(iterator, "iterator is null");
+        }
+
+        @Override
+        public Slice getNext()
+        {
+            if (!iterator.hasNext()) {
+                return null;
+            }
+            return iterator.next();
+        }
+
+        @Override
+        public boolean isFinished()
+        {
+            return !iterator.hasNext();
+        }
+
+        @Override
+        public ListenableFuture<Void> isBlocked()
+        {
+            return immediateVoidFuture();
+        }
+
+        @Override
+        public long getRetainedSizeInBytes()
+        {
+            return 0;
+        }
+
+        @Override
+        public void close()
+        {
+        }
+    }
+
+    @NotThreadSafe
+    private static class ExchangeOutputSource
+            implements OutputSource
+    {
+        private final Set<TaskId> selectedTasks;
+        private final ExchangeManager exchangeManager;
+        private final Exchange exchange;
+        private final QueryId queryId;
+
+        private ExchangeSource exchangeSource;
+        private boolean finished;
+
+        private ExchangeOutputSource(
+                Set<TaskId> selectedTasks,
+                ExchangeManager exchangeManager,
+                Exchange exchange,
+                QueryId queryId)
+        {
+            this.selectedTasks = ImmutableSet.copyOf(requireNonNull(selectedTasks, "selectedTasks is null"));
+            this.exchangeManager = requireNonNull(exchangeManager, "exchangeManager is null");
+            this.exchange = requireNonNull(exchange, "exchange is null");
+            this.queryId = requireNonNull(queryId, "queryId is null");
+        }
+
+        @Override
+        public Slice getNext()
+        {
+            if (finished) {
+                return null;
+            }
+            if (exchangeSource == null) {
+                CompletableFuture<List<ExchangeSourceHandle>> sourceHandlesFuture = exchange.getSourceHandles();
+                if (!sourceHandlesFuture.isDone()) {
+                    return null;
```

Comment:
The order of invokation of `isBlocked` and `poolPage` is inherently racy. It is possible that the `isBlocked` return "not blocked" (because of the `outputReady` feature) but then it become blocked on the `sourceHandlesFuture`.

Reply:
Sorry I don't understand here, the sink only has one partition and we called `addSink`, `noMoreSinks` as well as `sinkFinished` before this point, then `sourceHandlesFuture` must have finished.

Reply:
I don't fully understand. When `isBlocked` returns "not blocked", then `sourceHandlesFuture` must have been unblocked, otherwise `isBlocked` won't return "not blocked". 

When `outputReady` is done, you return `outputSource.isBlocked()`

Reply:
The contract between `getNext()` and `isBlocked()` invocation order is weak. The implementation usually shouldn't assume that the `isBlocked` is always in a not blocked state when the `getNext` is invoked.

The contract is weak as the engine is allowed to call `isBlocked` and `getNext` from multiple threads. When more than a single thread calls `isBlocked` and receives a signal that the  buffer is not blocked both threads may decide to call `getNext` but it is possible that the buffer will have data only for one of them. 

Under these circumstances it is hard to rely on strong guarantees. For example engine may sometime allow "spurious" wakeups and call `getNext` even when `isBlocked` hasn't been resolved. Thus it doesn't feel safe to assert here.